// @graphql-markdown/printer-legacy/dist/const/options.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DEFAULT_OPTIONS = exports.PRINT_TYPE_DEFAULT_OPTIONS = exports.SectionLevels = exports.TypeHierarchy = void 0;
const mdx_1 = __importDefault(require("../mdx"));
var TypeHierarchy;
(function (TypeHierarchy) {
    TypeHierarchy["API"] = "api";
    TypeHierarchy["ENTITY"] = "entity";
    TypeHierarchy["FLAT"] = "flat";
})(TypeHierarchy || (exports.TypeHierarchy = TypeHierarchy = {}));
var SectionLevels;
(function (SectionLevels) {
    SectionLevels["NONE"] = "";
    SectionLevels["LEVEL"] = "#";
})(SectionLevels || (exports.SectionLevels = SectionLevels = {}));
exports.PRINT_TYPE_DEFAULT_OPTIONS = {
    codeSection: true,
    deprecated: "default",
    exampleSection: false,
    metatags: [],
    parentTypePrefix: true,
    relatedTypeSection: true,
    typeBadges: true,
    hierarchy: { [TypeHierarchy.API]: {} },
};
exports.DEFAULT_OPTIONS = {
    ...exports.PRINT_TYPE_DEFAULT_OPTIONS,
    basePath: "/",
    collapsible: undefined,
    customDirectives: {},
    groups: undefined,
    frontMatter: {},
    level: undefined,
    metatags: [],
    onlyDocDirectives: [],
    parentType: undefined,
    schema: undefined,
    skipDocDirectives: [],
    withAttributes: false,
    ...mdx_1.default,
};
