// @graphql-markdown/printer-legacy/dist/const/strings.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MARKDOWN_SOC = exports.MARKDOWN_EOC = exports.MARKDOWN_EOP = exports.MARKDOWN_EOL = exports.MARKDOWN_CODE_INDENTATION = exports.FRONT_MATTER_DELIMITER = exports.MARKDOWN_CODE_SNIPPET = exports.NO_DESCRIPTION_TEXT = exports.GRAPHQL = exports.NON_NULL = exports.DEPRECATED = exports.ROOT_TYPE_LOCALE = void 0;
exports.ROOT_TYPE_LOCALE = {
    DIRECTIVE: { singular: "directive", plural: "directives" },
    ENUM: { singular: "enum", plural: "enums" },
    INPUT: { singular: "input", plural: "inputs" },
    INTERFACE: { singular: "interface", plural: "interfaces" },
    MUTATION: { singular: "mutation", plural: "mutations" },
    OPERATION: { singular: "operation", plural: "operations" },
    QUERY: { singular: "query", plural: "queries" },
    SCALAR: { singular: "scalar", plural: "scalars" },
    SUBSCRIPTION: { singular: "subscription", plural: "subscriptions" },
    TYPE: { singular: "object", plural: "objects" },
    UNION: { singular: "union", plural: "unions" },
};
exports.DEPRECATED = "deprecated";
exports.NON_NULL = "non-null";
exports.GRAPHQL = "graphql";
exports.NO_DESCRIPTION_TEXT = "No description";
exports.MARKDOWN_CODE_SNIPPET = "```";
exports.FRONT_MATTER_DELIMITER = "---";
exports.MARKDOWN_CODE_INDENTATION = "  ";
exports.MARKDOWN_EOL = "\n";
exports.MARKDOWN_EOP = `${exports.MARKDOWN_EOL.repeat(2)}`;
exports.MARKDOWN_EOC = `${exports.MARKDOWN_EOL}${exports.MARKDOWN_CODE_SNIPPET}${exports.MARKDOWN_EOL}`;
exports.MARKDOWN_SOC = `${exports.MARKDOWN_EOL}${exports.MARKDOWN_CODE_SNIPPET}${exports.GRAPHQL}${exports.MARKDOWN_EOL}`;
